File: app/src/main/java/com/oguz/spy/domain/TimerManager.kt
package com.oguz.spy.domain

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class TimerManager {

    fun startCountdown(durationInSeconds: Int): Flow<TimerState> = flow {
        var timeLeft = durationInSeconds

        while (timeLeft > 0) {
            val progress = timeLeft.toFloat() / durationInSeconds
            val minutes = timeLeft / 60
            val seconds = timeLeft % 60
            val formattedTime = String.format("%02d:%02d", minutes, seconds)

            val warningLevel = when {
                timeLeft <= 10 -> WarningLevel.CRITICAL
                timeLeft <= 30 -> WarningLevel.WARNING
                else -> WarningLevel.NORMAL
            }

            emit(
                TimerState(
                    timeLeft = timeLeft,
                    formattedTime = formattedTime,
                    progress = progress,
                    isFinished = false,
                    warningLevel = warningLevel
                )
            )

            delay(1000L)
            timeLeft--
        }

        emit(
            TimerState(
                timeLeft = 0,
                formattedTime = "00:00",
                progress = 0f,
                isFinished = true,
                warningLevel = WarningLevel.FINISHED
            )
        )
    }

    fun formatTime(totalSeconds: Int): String {
        val minutes = totalSeconds / 60
        val seconds = totalSeconds % 60
        return String.format("%02d:%02d", minutes, seconds)
    }

    fun getTimeWarningLevel(timeLeft: Int): WarningLevel {
        return when {
            timeLeft <= 0 -> WarningLevel.FINISHED
            timeLeft <= 10 -> WarningLevel.CRITICAL
            timeLeft <= 30 -> WarningLevel.WARNING
            else -> WarningLevel.NORMAL
        }
    }

    data class TimerState(
        val timeLeft: Int,
        val formattedTime: String,
        val progress: Float,
        val isFinished: Boolean,
        val warningLevel: WarningLevel
    )

    enum class WarningLevel {
        NORMAL,
        WARNING,
        CRITICAL,
        FINISHED
    }
}

File: lib/domain/logic/timer_manager.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'timer_manager.freezed.dart';

// ── Warning level ─────────────────────────────────────────────────────────────

enum WarningLevel { normal, warning, critical, finished }

// ── Timer state (freezed) ─────────────────────────────────────────────────────

@freezed
class TimerState with _$TimerState {
  const factory TimerState({
    required int timeLeft,
    required String formattedTime,
    /// 1.0 = full, 0.0 = done
    required double progress,
    required bool isFinished,
    required WarningLevel warningLevel,
  }) = _TimerState;
}

// ── Manager ───────────────────────────────────────────────────────────────────

/// Pure Dart — direct port of TimerManager.kt (Kotlin Flow → Dart Stream).
class TimerManager {
  /// Emits one [TimerState] per second then a final finished state.
  /// Mirrors Kotlin startCountdown() / `Flow<TimerState>`.
  Stream<TimerState> startCountdown(int durationInSeconds) async* {
    var timeLeft = durationInSeconds;

    while (timeLeft > 0) {
      yield _build(timeLeft, durationInSeconds);
      await Future<void>.delayed(const Duration(seconds: 1));
      timeLeft--;
    }

    yield TimerState(
      timeLeft: 0,
      formattedTime: '00:00',
      progress: 0.0,
      isFinished: true,
      warningLevel: WarningLevel.finished,
    );
  }

  /// Formats total seconds as MM:SS — mirrors Kotlin formatTime().
  String formatTime(int totalSeconds) {
    final m = (totalSeconds ~/ 60).toString().padLeft(2, '0');
    final s = (totalSeconds % 60).toString().padLeft(2, '0');
    return '$m:$s';
  }

  /// Mirrors Kotlin getTimeWarningLevel().
  WarningLevel warningLevelFor(int timeLeft) {
    if (timeLeft <= 0) return WarningLevel.finished;
    if (timeLeft <= 10) return WarningLevel.critical;
    if (timeLeft <= 30) return WarningLevel.warning;
    return WarningLevel.normal;
  }

  TimerState _build(int timeLeft, int total) => TimerState(
        timeLeft: timeLeft,
        formattedTime: formatTime(timeLeft),
        progress: total > 0 ? timeLeft / total : 0.0,
        isFinished: false,
        warningLevel: warningLevelFor(timeLeft),
      );
}
